src/edu/stanford/nlp/sempre/Learner.java

```diff
@@ -108,6 +108,9 @@ public void learn(int numIters, Map<String, List<Evaluation>> evaluations) {
       for (String group : dataset.groups()) {
         boolean lastIter = (iter == numIters);
         boolean updateWeights = opts.updateWeights && group.equals("train") && !lastIter;  // Don't train on last iteration
+        if (parser instanceof MutatingParser) {
```

Comment:
If you're going to do this, might as well make the mutate function part of Parser an overridable function...